File: integ/mongodb.py
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import asyncio
import os

from cases import run_not_found
from motor.motor_asyncio import AsyncIOMotorClient

from mirage import MountMode, Workspace
from mirage.resource.mongodb import MongoDBConfig, MongoDBResource
from mirage.types import CommandSafeguard

MONGODB_URI = os.environ.get("MONGODB_URI", "mongodb://localhost:27017")
DB = "mirage_integ"
MOUNT = "/mongodb"

BOOKS = [
    {
        "_id": 1,
        "title": "alpha",
        "author": "ada",
        "year": 2020,
        "tags": ["fiction", "classic"],
        "rating": 4.5,
    },
    {
        "_id": 2,
        "title": "beta",
        "author": "ben",
        "year": 2021,
        "tags": ["fiction"],
        "rating": 3.2,
    },
    {
        "_id": 3,
        "title": "gamma",
        "author": "cara",
        "year": 2022,
        "rating": 5.0,
    },
    {
        "_id": 4,
        "title": "delta",
        "author": "ada",
        "year": 2023,
        "tags": ["history"],
        "rating": 4.0,
    },
    {
        "_id": 5,
        "title": "epsilon",
        "author": "ben",
        "year": 2024,
        "rating": 2.5,
    },
]

AUTHORS = [
    {
        "_id": 1,
        "name": "ada",
        "books": 2
    },
    {
        "_id": 2,
        "name": "ben",
        "books": 2
    },
    {
        "_id": 3,
        "name": "cara",
        "books": 1
    },
]

CASES: list[tuple[str, str]] = [
    ("ls_root", f"ls {MOUNT}/"),
    ("ls_db", f"ls {MOUNT}/{DB}/"),
    ("ls_collections", f"ls {MOUNT}/{DB}/collections/"),
    ("ls_views", f"ls {MOUNT}/{DB}/views/"),
    ("ls_entity", f"ls {MOUNT}/{DB}/collections/books/"),
    ("tree", f"tree -L 3 {MOUNT}/{DB}/"),
    ("cat_database_json", f"cat {MOUNT}/{DB}/database.json"),
    ("cat_schema_json", f"cat {MOUNT}/{DB}/collections/books/schema.json"),
    ("cat_docs", f"cat {MOUNT}/{DB}/collections/books/documents.jsonl"),
    ("head_2", f"head -n 2 {MOUNT}/{DB}/collections/books/documents.jsonl"),
    ("tail_2", f"tail -n 2 {MOUNT}/{DB}/collections/books/documents.jsonl"),
    ("wc_l_books", f"wc -l {MOUNT}/{DB}/collections/books/documents.jsonl"),
    ("wc_default_books", f"wc {MOUNT}/{DB}/collections/books/documents.jsonl"),
    ("wc_l_authors",
     f"wc -l {MOUNT}/{DB}/collections/authors/documents.jsonl"),
    ("stat_docs", f"stat {MOUNT}/{DB}/collections/books/documents.jsonl"),
    ("grep_c_title",
     f"grep -c title {MOUNT}/{DB}/collections/books/documents.jsonl"),
    ("grep_ada", f"grep ada {MOUNT}/{DB}/collections/books/documents.jsonl"),
    ("grep_e_multi",
     f"grep -n -e ada -e ben {MOUNT}/{DB}/collections/books/documents.jsonl"),
    ("grep_db_scope", f"grep ada {MOUNT}/{DB}/"),
    ("grep_root_scope", f"grep ada {MOUNT}/"),
    ("rg_db_scope", f"rg ben {MOUNT}/{DB}/"),
    ("find_docs", f"find {MOUNT}/{DB}/ -name documents.jsonl"),
    ("find_schema", f"find {MOUNT}/{DB}/ -name schema.json"),
    ("jq_titles",
     f"jq '.title' {MOUNT}/{DB}/collections/books/documents.jsonl"),
    ("pipe_grep_c",
     f"cat {MOUNT}/{DB}/collections/books/documents.jsonl | grep -c fiction"),
    ("cat_view_docs", f"cat {MOUNT}/{DB}/views/recent_books/documents.jsonl"),
    ("wc_l_view", f"wc -l {MOUNT}/{DB}/views/recent_books/documents.jsonl"),

    # ----- safeguard: per-mount cap on cat (set to 2 lines below) -----
    ("safeguard_cat_truncates",
     f"cat {MOUNT}/{DB}/collections/books/documents.jsonl"),
    ("safeguard_cat_pipe_uncapped",
     f"cat {MOUNT}/{DB}/collections/books/documents.jsonl | wc -l"),
]


async def _seed(client: AsyncIOMotorClient) -> None:
    await client.drop_database(DB)
    db = client[DB]
    await db["books"].insert_many([dict(d) for d in BOOKS])
    await db["authors"].insert_many([dict(d) for d in AUTHORS])
    await db.create_collection(
        "recent_books",
        viewOn="books",
        pipeline=[{
            "$match": {
                "year": {
                    "$gte": 2022
                }
            }
        }],
    )


async def _run(ws: Workspace, name: str, cmd: str) -> None:
    result = await ws.execute(cmd)
    out = await result.stdout_str()
    print(f"=== {name} ===")
    print(out, end="" if out.endswith("\n") else "\n")
    if name.startswith("safeguard_"):
        err = await result.stderr_str()
        if err:
            print(err, end="" if err.endswith("\n") else "\n")


def _set_cat_safeguard(ws: Workspace, max_lines: int) -> None:
    sg = CommandSafeguard(max_lines=max_lines)
    mounts = list(ws._registry._mounts)
    if ws._registry.default_mount is not None:
        mounts.append(ws._registry.default_mount)
    for m in mounts:
        m.command_safeguards["cat"] = sg


async def main() -> None:
    seed_client = AsyncIOMotorClient(MONGODB_URI)
    try:
        await _seed(seed_client)
    finally:
        seed_client.close()
    resource = MongoDBResource(
        config=MongoDBConfig(uri=MONGODB_URI, databases=[DB]))
    ws = Workspace({MOUNT: resource}, mode=MountMode.READ)
    for name, cmd in CASES:
        if name == "safeguard_cat_truncates":
            _set_cat_safeguard(ws, max_lines=2)
        await _run(ws, name, cmd)
    await run_not_found(ws, MOUNT)


if __name__ == "__main__":
    asyncio.run(main())

File: integ/mongodb.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import { Double, MongoClient } from "mongodb";
import {
  CommandSafeguard,
  MongoDBResource,
  MountMode,
  Workspace,
} from "@struktoai/mirage-node";
import { runNotFound } from "./cases.ts";

const MONGODB_URI = process.env.MONGODB_URI ?? "mongodb://localhost:27017";
const DB = "mirage_integ";
const MOUNT = "/mongodb";

const DEC = new TextDecoder();

const BOOKS = [
  { _id: 1, title: "alpha", author: "ada", year: 2020, tags: ["fiction", "classic"], rating: 4.5 },
  { _id: 2, title: "beta", author: "ben", year: 2021, tags: ["fiction"], rating: 3.2 },
  { _id: 3, title: "gamma", author: "cara", year: 2022, rating: 5.0 },
  { _id: 4, title: "delta", author: "ada", year: 2023, tags: ["history"], rating: 4.0 },
  { _id: 5, title: "epsilon", author: "ben", year: 2024, rating: 2.5 },
];

const AUTHORS = [
  { _id: 1, name: "ada", books: 2 },
  { _id: 2, name: "ben", books: 2 },
  { _id: 3, name: "cara", books: 1 },
];

const CASES: ReadonlyArray<readonly [string, string]> = [
  ["ls_root", `ls ${MOUNT}/`],
  ["ls_db", `ls ${MOUNT}/${DB}/`],
  ["ls_collections", `ls ${MOUNT}/${DB}/collections/`],
  ["ls_views", `ls ${MOUNT}/${DB}/views/`],
  ["ls_entity", `ls ${MOUNT}/${DB}/collections/books/`],
  ["tree", `tree -L 3 ${MOUNT}/${DB}/`],
  ["cat_database_json", `cat ${MOUNT}/${DB}/database.json`],
  ["cat_schema_json", `cat ${MOUNT}/${DB}/collections/books/schema.json`],
  ["cat_docs", `cat ${MOUNT}/${DB}/collections/books/documents.jsonl`],
  ["head_2", `head -n 2 ${MOUNT}/${DB}/collections/books/documents.jsonl`],
  ["tail_2", `tail -n 2 ${MOUNT}/${DB}/collections/books/documents.jsonl`],
  ["wc_l_books", `wc -l ${MOUNT}/${DB}/collections/books/documents.jsonl`],
  ["wc_default_books", `wc ${MOUNT}/${DB}/collections/books/documents.jsonl`],
  ["wc_l_authors", `wc -l ${MOUNT}/${DB}/collections/authors/documents.jsonl`],
  ["stat_docs", `stat ${MOUNT}/${DB}/collections/books/documents.jsonl`],
  ["grep_c_title", `grep -c title ${MOUNT}/${DB}/collections/books/documents.jsonl`],
  ["grep_ada", `grep ada ${MOUNT}/${DB}/collections/books/documents.jsonl`],
  ["grep_e_multi", `grep -n -e ada -e ben ${MOUNT}/${DB}/collections/books/documents.jsonl`],
  ["grep_db_scope", `grep ada ${MOUNT}/${DB}/`],
  ["grep_root_scope", `grep ada ${MOUNT}/`],
  ["rg_db_scope", `rg ben ${MOUNT}/${DB}/`],
  ["find_docs", `find ${MOUNT}/${DB}/ -name documents.jsonl`],
  ["find_schema", `find ${MOUNT}/${DB}/ -name schema.json`],
  ["jq_titles", `jq '.title' ${MOUNT}/${DB}/collections/books/documents.jsonl`],
  ["pipe_grep_c", `cat ${MOUNT}/${DB}/collections/books/documents.jsonl | grep -c fiction`],
  ["cat_view_docs", `cat ${MOUNT}/${DB}/views/recent_books/documents.jsonl`],
  ["wc_l_view", `wc -l ${MOUNT}/${DB}/views/recent_books/documents.jsonl`],
  ["safeguard_cat_truncates", `cat ${MOUNT}/${DB}/collections/books/documents.jsonl`],
  ["safeguard_cat_pipe_uncapped", `cat ${MOUNT}/${DB}/collections/books/documents.jsonl | wc -l`],
];

async function seed(client: MongoClient): Promise<void> {
  const db = client.db(DB);
  await db.dropDatabase();
  // Python seeds floats (BSON double); insert Double so the inferred schema
  // matches (JS whole numbers would otherwise serialize to BSON int32).
  await db
    .collection("books")
    .insertMany(BOOKS.map((d) => ({ ...d, rating: new Double(d.rating) })));
  await db.collection("authors").insertMany(AUTHORS.map((d) => ({ ...d })));
  await db.createCollection("recent_books", {
    viewOn: "books",
    pipeline: [{ $match: { year: { $gte: 2022 } } }],
  });
}

async function run(ws: Workspace, name: string, cmd: string): Promise<void> {
  const result = await ws.execute(cmd);
  const out = DEC.decode(result.stdout);
  process.stdout.write(`=== ${name} ===\n`);
  process.stdout.write(out.endsWith("\n") ? out : out + "\n");
  if (name.startsWith("safeguard_")) {
    const err = DEC.decode(result.stderr);
    if (err) process.stdout.write(err.endsWith("\n") ? err : err + "\n");
  }
}

function setCatSafeguard(ws: Workspace, maxLines: number): void {
  const sg = new CommandSafeguard({ maxLines });
  for (const m of ws.registry.allMounts()) m.commandSafeguards.set("cat", sg);
  if (ws.registry.defaultMount !== null) {
    ws.registry.defaultMount.commandSafeguards.set("cat", sg);
  }
}

async function main(): Promise<void> {
  const seedClient = new MongoClient(MONGODB_URI);
  await seedClient.connect();
  try {
    await seed(seedClient);
  } finally {
    await seedClient.close();
  }
  const resource = new MongoDBResource({ uri: MONGODB_URI, databases: [DB] });
  const ws = new Workspace({ [MOUNT]: resource }, { mode: MountMode.READ });
  try {
    for (const [name, cmd] of CASES) {
      if (name === "safeguard_cat_truncates") setCatSafeguard(ws, 2);
      await run(ws, name, cmd);
    }
    await runNotFound(ws, MOUNT);
  } finally {
    await ws.close();
    await resource.close();
  }
}

main().catch((err: unknown) => {
  console.error(err);
  process.exit(1);
});
